docs/corfu-queue-design.md

```diff
@@ -0,0 +1,12 @@
+The class `CorfuQueue` implements a persisted queue over the abstraction of a CorfuTable. Currently CorfuTable uses a HashMap<> to represent the materialized view of a distributed map. A HashMap does not carry a notion of ordering natively. However when implemented over the abstraction of a distributed shared log, the elements added to the map do, in fact, have an ordering imposed by their append operations into the global log. The class `CorfuQueue` merely exposes this inherent ordering as an immutable logical FIFO Queue interface with three simple apis:
+
+###  1. `String enqueue(E)`
+Since a map has a key and value, where key is a conflict parameter, enqueue simply generates a non-conflicting UUID as the key and inserts the Entry as a value into a CorfuTable. The generated UUID is returned for the caller.
```

Comment:
How do we plan on generating non-conflicting UUIDs? It either that we will need something similar to Twitter Snowflake, or rely on the sequencer for the uniqueness.

Reply:
That is a good question. For the prototype I was thinking of using Java UUID which as a collision probability of 1 in a billion
https://en.wikipedia.org/wiki/Universally_unique_identifier#Collisions
Since in our system we already have a sequencer which issues tokens and tokens also need to have the same property of uniqueness, I was hoping we can use tokens without breaking the abstraction.

Reply:
Now that I think about it, even conflicting UUIDs should be fine, as along as you are able to detect the collision and re-generate the UUID.

Reply:
Its actually not fine. Collision means data loss in this case. 

Consider this case: two clients successfully enqueue two items, which generates two UUIDs u1 and u2 for write1 and write2. Those IDs are then published, on the dequeue (i.e. the consumer), two different consumers cannot map u1 -> write1 and u2 -> write2 since u1 and u2 are equal, so they must be unique on generation.